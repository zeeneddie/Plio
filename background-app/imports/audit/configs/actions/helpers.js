import { _ } from 'meteor/underscore';
import { without } from 'ramda';

import {
  Standards,
  NonConformities,
  Risks,
  Goals,
} from '../../../share/collections';
import { ProblemTypes, DocumentTypes } from '../../../share/constants';
import { getUserId, getLinkedDocAuditConfig } from '../../utils/helpers';

export { getLinkedDocName, getLinkedDocDescription } from '../../utils/helpers';

export { getLinkedDocAuditConfig };

export const getReceivers = ({ linkedTo, notify }, user) => {
  const getLinkedDocsIds = (linkedDocs, docType) => _.pluck(
    linkedDocs.filter(({ documentType }) => documentType === docType),
    'documentId',
  );

  const usersIds = new Set(notify);
  const standardsIds = new Set();

  const getIds = (collection, problemType) => {
    const query = {
      _id: { $in: getLinkedDocsIds(linkedTo, problemType) },
    };
    const options = {
      fields: {
        standardsIds: 1,
        originatorId: 1,
      },
    };

    collection.find(query, options).forEach((doc) => {
      doc.standardsIds.forEach(id => standardsIds.add(id));
      usersIds.add(doc.originatorId);
    });
  };

  const problemCollections = [
    { collection: NonConformities, type: ProblemTypes.NON_CONFORMITY },
    { collection: NonConformities, type: ProblemTypes.POTENTIAL_GAIN },
    { collection: Risks, type: ProblemTypes.RISK },
  ];

  problemCollections.forEach(({ collection, type }) => getIds(collection, type));

  Standards.find({
    _id: { $in: Array.from(standardsIds) },
  }, { fields: { owner: 1 } }).forEach(({ owner }) => usersIds.add(owner));

  Goals.find({
    _id: {
      $in: getLinkedDocsIds(linkedTo, DocumentTypes.GOAL),
    },
  }, { fields: { ownerId: 1 } }).forEach(({ ownerId }) => usersIds.add(ownerId));

  const receivers = Array.from(usersIds);

  return without(getUserId(user), receivers);
};
