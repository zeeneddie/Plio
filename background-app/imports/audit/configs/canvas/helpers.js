export { getNotifyReceivers as getReceivers } from '../../utils/helpers';

export function emailTemplateData({ newDoc, auditConfig }) {
  return {
    button: {
      label: 'Go to this canvas',
      url: auditConfig.docUrl(newDoc),
    },
  };
}
