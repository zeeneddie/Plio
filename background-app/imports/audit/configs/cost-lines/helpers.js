export { getNotifyReceivers as getReceivers } from '../../utils/helpers';
