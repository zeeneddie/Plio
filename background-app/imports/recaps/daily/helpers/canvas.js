import pluralize from 'pluralize';

export const getDailyRecapCanvasTitle = ({ count, description }) =>
  `${count} ${pluralize(description, count)} ${pluralize('was', count)} ` +
  'updated in the business model canvas:';
