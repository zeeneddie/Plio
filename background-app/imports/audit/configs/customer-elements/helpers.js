import { getNotifyReceivers, getLinkedDocAuditConfig } from '../../utils/helpers';
import { getCollectionByDocType } from '../../../share/helpers';

export { emailTemplateData } from '../canvas/helpers';

export const getReceivers = (customerElement, user) =>
  customerElement.linkedTo.map(({ documentId, documentType }) => {
    const doc = getCollectionByDocType(documentType).findOne({ _id: documentId });
    const linkedAuditConfig = getLinkedDocAuditConfig(documentType);
    const notify = linkedAuditConfig.docNotifyList(doc);

    return getNotifyReceivers({ notify }, user);
  });
