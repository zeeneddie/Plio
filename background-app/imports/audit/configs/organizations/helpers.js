
import { getUserId } from '../../utils/helpers';
import { getOwnerId } from '../../../share/helpers/Organization';

export const getReceivers = function (doc, user) {
  const executorId = getUserId(user);
  const ownerId = getOwnerId(doc);

  return executorId === ownerId ? [] : [ownerId];
};
