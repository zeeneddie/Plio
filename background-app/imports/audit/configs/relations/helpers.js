import { contains } from 'ramda';

import { CustomerElementTypes } from '../../../share/constants';
import { getLinkedDoc, getLinkedDocAuditConfig } from '../../utils/helpers';
import { getReceivers } from '../customer-elements/helpers';

export const isCustomerElement = ({ documentType }) =>
  contains(documentType, Object.values(CustomerElementTypes));

export const getCustomerElementData = ({ rel1, rel2 }) => {
  const doc1 = getLinkedDoc(rel1.documentId, rel1.documentType);
  const doc2 = getLinkedDoc(rel2.documentId, rel2.documentType);
  const rel1AuditConfig = getLinkedDocAuditConfig(rel1.documentType);
  const rel2AuditConfig = getLinkedDocAuditConfig(rel2.documentType);
  const rel1Name = rel1AuditConfig.docName(doc1);
  const rel1Desc = rel1AuditConfig.docDescription(doc1);
  const rel2Name = rel2AuditConfig.docName(doc2);
  const rel2Desc = rel2AuditConfig.docDescription(doc2);
  // WARN: Careful! Things will break if customer elements
  // could be linked to multiple canvas documents
  const { linkedTo: [rel1LinkedTo] } = doc1;
  const { linkedTo: [rel2LinkedTo] } = doc2;
  const rel1LinkedDoc = getLinkedDoc(rel1LinkedTo.documentId, rel1LinkedTo.documentType);
  const rel2LinkedDoc = getLinkedDoc(rel2LinkedTo.documentId, rel2LinkedTo.documentType);
  const rel1LinkedDocAuditConfig = getLinkedDocAuditConfig(rel1LinkedTo.documentType);
  const rel2LinkedDocAuditConfig = getLinkedDocAuditConfig(rel2LinkedTo.documentType);
  const rel1LinkedDocName = rel1LinkedDocAuditConfig.docName(rel1LinkedDoc);
  const rel1LinkedDocDesc = rel1LinkedDocAuditConfig.docDescription(rel1LinkedDoc);
  const rel2LinkedDocName = rel2LinkedDocAuditConfig.docName(rel2LinkedDoc);
  const rel2LinkedDocDesc = rel2LinkedDocAuditConfig.docDescription(rel2LinkedDoc);

  return {
    rel1Name,
    rel1Desc,
    rel2Name,
    rel2Desc,
    rel1LinkedDocDesc,
    rel2LinkedDocDesc,
    rel1LinkedDocName,
    rel2LinkedDocName,
  };
};

export const getCustomerElementReceivers = ({ rel1: { documentId, documentType } }, user) => {
  const doc = getLinkedDoc(documentId, documentType);
  return getReceivers(doc, user);
};
