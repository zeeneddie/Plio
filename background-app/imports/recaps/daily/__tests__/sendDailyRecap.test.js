import { __setupDB, __closeDB, __clearDB } from 'meteor/mongo';
import { Meteor } from 'meteor/meteor';
import faker from 'faker';
import moment from 'moment';
import { times, flatten } from 'ramda';

import createContext from '../../../share/utils/tests/createContext';
import { DEFAULT_ORG_TIMEZONE, CollectionNames } from '../../../share/constants';
import NotificationSender from '../../../share/utils/NotificationSender';
import { getPrettyTzDate } from '../../../helpers/date';

jest.mock('../../../share/utils/NotificationSender');

describe('sendDailyRecap', () => {
  faker.seed(123);

  let context;
  const timezone = DEFAULT_ORG_TIMEZONE;

  beforeAll(async () => {
    await __setupDB();

    context = createContext({});

    Meteor.users = context.collections.Users;
    jest.doMock('../../../share/collections', () => context.collections);
  });
  afterAll(__closeDB);
  beforeEach(async () => {
    await __clearDB();

    NotificationSender.mockClear();
    context.collections.Users.findOne.mockClear();
  });

  it('should default to yesterday if no date provided', async () => {
    const { sendDailyRecap } = require('../sendDailyRecap');

    const date = moment().subtract(1, 'day').toDate();
    const organizationId = await context.collections.Organizations.insert({
      timezone,
    });

    await Promise.all([
      context.collections.AuditLogs.insert({
        organizationId,
        date,
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.word(),
      }),
      context.collections.AuditLogs.insert({
        organizationId,
        date: moment().add(1, 'day').toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.word(),
      }),
      context.collections.AuditLogs.insert({
        organizationId,
        date: moment().subtract(2, 'days').toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.word(),
      }),
    ]);

    await sendDailyRecap(organizationId);

    const expectedDate = getPrettyTzDate(date, timezone, 'h:mm A');

    expect(NotificationSender.mock.calls[0][0].templateData.recapData[0].data).toHaveLength(1);
    expect(NotificationSender.mock.calls[0][0].templateData.recapData[0].data[0].logs[0].date)
      .toBe(expectedDate);
  });

  it('should use provided date for daily recap', async () => {
    const { sendDailyRecap } = require('../sendDailyRecap');

    const date = new Date(2018, 0, 1);
    const organizationId = await context.collections.Organizations.insert({
      timezone,
    });

    await Promise.all([
      context.collections.AuditLogs.insert({
        organizationId,
        date,
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.word(),
      }),
      context.collections.AuditLogs.insert({
        organizationId,
        date: moment(date).add(1, 'day').toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.word(),
      }),
      context.collections.AuditLogs.insert({
        organizationId,
        date: moment(date).subtract(2, 'days').toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.word(),
      }),
    ]);

    await sendDailyRecap(organizationId, date);

    const expectedDate = getPrettyTzDate(date, timezone, 'h:mm A');

    expect(NotificationSender.mock.calls[0][0].templateData.recapData[0].data[0].logs[0].date)
      .toBe(expectedDate);
  });

  it('ensures proper recipients', async () => {
    const { sendDailyRecap } = require('../sendDailyRecap');

    const date = moment().subtract(1, 'day').toDate();
    const [, ...recipientIds] = await Promise.all([
      context.collections.Users.insert({}),
      context.collections.Users.insert({ _id: context.userId }),
      context.collections.Users.insert({}),
      context.collections.Users.insert({}),
    ]);
    const users = recipientIds.map((userId, i) => ({
      userId,
      sendDailyRecap: i !== 1,
      isRemoved: i === 2,
    }));
    const organizationId = faker.random.uuid();
    await context.collections.Organizations.addMembers({ _id: organizationId, timezone }, users);

    await context.collections.AuditLogs.insert({
      organizationId,
      date,
      executor: context.userId,
      collection: CollectionNames.ORGANIZATIONS,
      documentId: organizationId,
      message: faker.random.word(),
    });

    await sendDailyRecap(organizationId);

    expect(NotificationSender.mock.calls[0][0].recipients).toEqual([context.userId]);
  });

  it('queries logs over the past 24hrs', async () => {
    const { sendDailyRecap } = require('../sendDailyRecap');

    const organizationId = await context.collections.Organizations.insert({ timezone });
    await Promise.all([
      ...times(n => context.collections.AuditLogs.insert({
        organizationId,
        date: moment()
          .tz(timezone)
          .subtract(1, 'day')
          .startOf('day')
          .add(n + 1, 'hours')
          .toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.words(),
      }), 24),
      context.collections.AuditLogs.insert({
        organizationId,
        date: moment().subtract(10, 'days').toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.words(),
      }),
      context.collections.AuditLogs.insert({
        organizationId,
        date: moment().add(10, 'days').toDate(),
        executor: context.userId,
        collection: CollectionNames.ORGANIZATIONS,
        documentId: organizationId,
        message: faker.random.words(),
      }),
    ]);

    await sendDailyRecap(organizationId);

    expect(NotificationSender.mock.calls[0][0].templateData.recapData[0].data[0].logs)
      .toHaveLength(23);
  });

  it('shouldn\t send emails when there are no data', async () => {
    const { sendDailyRecap } = require('../sendDailyRecap');

    const promise = sendDailyRecap(faker.random.uuid());

    await expect(promise).rejects.toEqual(expect.any(Error));
    expect(NotificationSender).not.toHaveBeenCalled();

    const organizationId = faker.random.uuid();
    await context.collections.Organizations.addMembers(
      { _id: organizationId, timezone },
      [
        { userId: context.userId, sendDailyRecap: true },
      ],
    );

    await sendDailyRecap(organizationId);

    expect(NotificationSender).not.toHaveBeenCalled();
  });

  it('properly counts documents', async () => {
    const { sendDailyRecap } = require('../sendDailyRecap');

    const organizationId = await context.collections.Organizations.insert({
      timezone,
      serialNumber: faker.random.number(),
    });
    const [keyPartnerId, ...keyActivityIds] = await Promise.all([
      context.collections.KeyPartners.insert({ organizationId }),
      ...times(() => context.collections.KeyActivities.insert({ organizationId }), 3),
    ]);
    await Promise.all([
      ...times(() => [
        context.collections.AuditLogs.insert({
          organizationId,
          date: moment().subtract(1, 'day').toDate(),
          executor: context.userId,
          collection: CollectionNames.KEY_PARTNERS,
          documentId: keyPartnerId,
          message: faker.random.words(),
        }),
      ], 3),
      ...keyActivityIds.map(id => times(() => context.collections.AuditLogs.insert({
        organizationId,
        date: moment().subtract(1, 'day').toDate(),
        executor: context.userId,
        collection: CollectionNames.KEY_ACTIVITIES,
        documentId: id,
        message: faker.random.word(),
      }), 3)),
    ]);

    await sendDailyRecap(organizationId);

    const data = NotificationSender.mock.calls[0][0].templateData.recapData;

    expect(data[0].title).toEqual(expect.stringMatching(/3 .+ were updated(.+)?:/));

    expect(data[1].title).toEqual(expect.stringMatching(/1 .+ was updated(.+)?:/));
  });

  it('matches snapshot', async () => {
    const { sendDailyRecap, getRecapDate } = require('../sendDailyRecap');

    const {
      Channels,
      CostLines,
      CustomerRelationships,
      CustomerSegments,
      KeyActivities,
      KeyPartners,
      KeyResources,
      RevenueStreams,
      ValuePropositions,
      Actions,
      NonConformities,
      Risks,
      Standards,
      Goals,
      Milestones,
      Benefits,
      Features,
      Needs,
      Wants,
    } = context.collections;

    const date = moment('2018-01-01').add(5, 'hours').toDate();
    const organizationName = faker.random.word();
    const subject = `Daily recap for ${getPrettyTzDate(getRecapDate(date, timezone), timezone)}`;

    await context.collections.Users.insert({ _id: context.userId });
    const organizationId = await context.collections.Organizations.insert({
      timezone,
      name: organizationName,
      serialNumber: faker.random.number(),
    });
    await context.collections.Organizations.addMembers(
      { _id: organizationId },
      [{ userId: context.userId, sendDailyRecap: true }],
    );

    const collections = [
      Channels,
      CostLines,
      CustomerRelationships,
      CustomerSegments,
      KeyActivities,
      KeyPartners,
      KeyResources,
      RevenueStreams,
      ValuePropositions,
      Actions,
      NonConformities,
      Risks,
      Standards,
      Goals,
      Milestones,
      Benefits,
      Features,
      Needs,
      Wants,
    ];

    await context.collections.AuditLogs.insert({
      organizationId,
      date,
      executor: context.userId,
      collection: CollectionNames.ORGANIZATIONS,
      documentId: organizationId,
      message: faker.random.word(),
    });

    const ids = await Promise.all(
      collections.map(collection => collection.insert({
        title: faker.random.word(),
        sequentialId: faker.random.word(),
        serialNumber: faker.random.number(),
      })),
    );
    await Promise.all([
      ...flatten(times(() => ids.map((id, i) => context.collections.AuditLogs.insert({
        organizationId,
        date,
        executor: context.userId,
        collection: collections[i].collection.collectionName,
        documentId: id,
        message: faker.random.words(),
        field: faker.random.word(),
        newValue: faker.random.word(),
        oldValue: faker.random.word(),
      })), 2)),
      context.collections.WorkItems.insert({
        linkedDoc: {
          _id: (await context.collections.Actions.findOne({}))._id,
        },
      }),
    ]);

    await sendDailyRecap(organizationId, date);

    expect(NotificationSender).toHaveBeenCalledWith({
      templateName: 'recapEmail',
      recipients: [context.userId],
      emailSubject: subject,
      templateData: {
        organizationName,
        title: subject,
        unsubscribeUrl: expect.stringMatching('/unsubscribe'),
        recapData: expect.any(Array),
      },
    });

    expect(NotificationSender.mock.calls[0][0]).toMatchSnapshot();
  });
});
