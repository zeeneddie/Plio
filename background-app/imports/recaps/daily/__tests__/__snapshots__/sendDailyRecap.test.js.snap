// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendDailyRecap matches snapshot 1`] = `
Object {
  "emailSubject": "Daily recap for January 01, 2018",
  "recipients": Array [
    "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
  ],
  "templateData": Object {
    "organizationName": "Peru",
    "recapData": Array [
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Actions",
                "date": "3:00 AM",
                "documentId": "922d73f3-603c-4e39-8ba8-105cdf5b8ccd",
                "executor": "",
                "message": "Intuitive circuit",
              },
              Object {
                "collection": "Actions",
                "date": "3:00 AM",
                "documentId": "922d73f3-603c-4e39-8ba8-105cdf5b8ccd",
                "executor": "",
                "message": "Quality",
              },
            ],
            "title": "best-of-breed \\"Investor\\"",
            "url": "99173/work-inbox?id=ef3875c5-6e69-4c69-8689-3d35420f4a93",
          },
        ],
        "title": "1 action was updated:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Channels",
                "date": "3:00 AM",
                "documentId": "9d43762e-2b19-41b8-8c6c-5fe2f2d7bdec",
                "executor": "",
                "message": "flexibility",
              },
              Object {
                "collection": "Channels",
                "date": "3:00 AM",
                "documentId": "9d43762e-2b19-41b8-8c6c-5fe2f2d7bdec",
                "executor": "",
                "message": "communities",
              },
            ],
            "title": "incentivize",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Channel was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "CostLines",
                "date": "3:00 AM",
                "documentId": "61bc5205-9ab2-4ba2-85c6-66e116180b09",
                "executor": "",
                "message": "Sleek Steel Computer Iran",
              },
              Object {
                "collection": "CostLines",
                "date": "3:00 AM",
                "documentId": "61bc5205-9ab2-4ba2-85c6-66e116180b09",
                "executor": "",
                "message": "ivory transmitter",
              },
            ],
            "title": "Towels",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Cost Line was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "CustomerRelationships",
                "date": "3:00 AM",
                "documentId": "97faddae-e1f8-43fb-8e30-fca1ce6972d0",
                "executor": "",
                "message": "Awesome 3rd generation",
              },
              Object {
                "collection": "CustomerRelationships",
                "date": "3:00 AM",
                "documentId": "97faddae-e1f8-43fb-8e30-fca1ce6972d0",
                "executor": "",
                "message": "directional facilitate",
              },
            ],
            "title": "Tools",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Customer Relationship was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "CustomerSegments",
                "date": "3:00 AM",
                "documentId": "f814d36a-d7c8-4d37-8bc5-40d3d0154780",
                "executor": "",
                "message": "infomediaries Small Concrete Fish SCSI",
              },
              Object {
                "collection": "CustomerSegments",
                "date": "3:00 AM",
                "documentId": "f814d36a-d7c8-4d37-8bc5-40d3d0154780",
                "executor": "",
                "message": "Assistant Louisiana JBOD",
              },
            ],
            "title": "Bacon",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Customer Segment was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "KeyActivities",
                "date": "3:00 AM",
                "documentId": "181171bd-f347-4171-9331-b04442d0651d",
                "executor": "",
                "message": "SMS",
              },
              Object {
                "collection": "KeyActivities",
                "date": "3:00 AM",
                "documentId": "181171bd-f347-4171-9331-b04442d0651d",
                "executor": "",
                "message": "Checking Account Branding",
              },
            ],
            "title": "red",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Key Activity was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "KeyPartners",
                "date": "3:00 AM",
                "documentId": "bf07f52e-bc67-4d1f-a37d-481ffa0953ab",
                "executor": "",
                "message": "Small Cotton Mouse Concrete Ergonomic",
              },
              Object {
                "collection": "KeyPartners",
                "date": "3:00 AM",
                "documentId": "bf07f52e-bc67-4d1f-a37d-481ffa0953ab",
                "executor": "",
                "message": "monitor",
              },
            ],
            "title": "Grove",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Key Partner was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "KeyResources",
                "date": "3:00 AM",
                "documentId": "ff04f552-70e2-4c03-81bd-0bad368727b5",
                "executor": "",
                "message": "withdrawal indigo EXE",
              },
              Object {
                "collection": "KeyResources",
                "date": "3:00 AM",
                "documentId": "ff04f552-70e2-4c03-81bd-0bad368727b5",
                "executor": "",
                "message": "Legacy system yellow",
              },
            ],
            "title": "copy",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Key Resource was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "NonConformities",
                "date": "3:00 AM",
                "documentId": "74872e70-45d9-4581-800d-bd6ed196063d",
                "executor": "",
                "message": "Accountability British Indian Ocean Territory (Chagos Archipelago)",
              },
              Object {
                "collection": "NonConformities",
                "date": "3:00 AM",
                "documentId": "74872e70-45d9-4581-800d-bd6ed196063d",
                "executor": "",
                "message": "payment Solomon Islands Dollar Rustic",
              },
            ],
            "title": "discrete \\"Strategist\\"",
            "url": "99173/non-conformities/74872e70-45d9-4581-800d-bd6ed196063d",
          },
        ],
        "title": "1 nonconformity was updated:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "RevenueStreams",
                "date": "3:00 AM",
                "documentId": "cad6dec8-5ab3-4f8e-af60-1993e35d6cd5",
                "executor": "",
                "message": "SCSI",
              },
              Object {
                "collection": "RevenueStreams",
                "date": "3:00 AM",
                "documentId": "cad6dec8-5ab3-4f8e-af60-1993e35d6cd5",
                "executor": "",
                "message": "incremental",
              },
            ],
            "title": "Kentucky",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Revenue Stream was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Risks",
                "date": "3:00 AM",
                "documentId": "b523afb3-2237-4b4c-afd4-f82276a405bc",
                "executor": "",
                "message": "Intelligent Games Home",
              },
              Object {
                "collection": "Risks",
                "date": "3:00 AM",
                "documentId": "b523afb3-2237-4b4c-afd4-f82276a405bc",
                "executor": "",
                "message": "Fresh Home",
              },
            ],
            "title": "Plastic \\"Indiana\\"",
            "url": "99173/risks/b523afb3-2237-4b4c-afd4-f82276a405bc",
          },
        ],
        "title": "1 risk was updated:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Standards",
                "date": "3:00 AM",
                "documentId": "114cf37d-9b0c-48eb-8011-151f63c0aced",
                "executor": "",
                "message": "invoice Associate projection",
              },
              Object {
                "collection": "Standards",
                "date": "3:00 AM",
                "documentId": "114cf37d-9b0c-48eb-8011-151f63c0aced",
                "executor": "",
                "message": "Shoes Turks and Caicos Islands Pants",
              },
            ],
            "title": "\\"schemas\\"",
            "url": "99173/standards/114cf37d-9b0c-48eb-8011-151f63c0aced",
          },
        ],
        "title": "1 standard was updated:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "ValuePropositions",
                "date": "3:00 AM",
                "documentId": "09e6bc87-0667-4682-92e2-79a1133f331c",
                "executor": "",
                "message": "Sports Switchable Compatible",
              },
              Object {
                "collection": "ValuePropositions",
                "date": "3:00 AM",
                "documentId": "09e6bc87-0667-4682-92e2-79a1133f331c",
                "executor": "",
                "message": "Awesome Frozen Keyboard",
              },
            ],
            "title": "payment",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Value Proposition was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Goals",
                "date": "3:00 AM",
                "documentId": "2b480f94-ea93-416b-8a52-0d1e7654395f",
                "executor": "",
                "message": "Cross-group Ghana",
              },
              Object {
                "collection": "Goals",
                "date": "3:00 AM",
                "documentId": "2b480f94-ea93-416b-8a52-0d1e7654395f",
                "executor": "",
                "message": "teal front-end",
              },
            ],
            "title": "Pataca",
            "url": "99173",
          },
        ],
        "title": "1 key goal was updated:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Milestones",
                "date": "3:00 AM",
                "documentId": "c6fada47-3a35-4373-beff-f101ad790397",
                "executor": "",
                "message": "navigating Corporate deposit",
              },
              Object {
                "collection": "Milestones",
                "date": "3:00 AM",
                "documentId": "c6fada47-3a35-4373-beff-f101ad790397",
                "executor": "",
                "message": "deposit Accountability",
              },
            ],
            "title": "Assurance",
            "url": "99173",
          },
        ],
        "title": "1 milestone was updated:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Benefits",
                "date": "3:00 AM",
                "documentId": "23c2f4b2-1999-420d-badf-f247c8814a09",
                "executor": "",
                "message": "e-enable",
              },
              Object {
                "collection": "Benefits",
                "date": "3:00 AM",
                "documentId": "23c2f4b2-1999-420d-badf-f247c8814a09",
                "executor": "",
                "message": "product Mouse",
              },
            ],
            "title": "BE65843 \\"approach\\"",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Benefit was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Features",
                "date": "3:00 AM",
                "documentId": "c45ffee6-12db-4ac0-bf4d-ddbe7e23fec0",
                "executor": "",
                "message": "Court Money Market Account",
              },
              Object {
                "collection": "Features",
                "date": "3:00 AM",
                "documentId": "c45ffee6-12db-4ac0-bf4d-ddbe7e23fec0",
                "executor": "",
                "message": "Computers",
              },
            ],
            "title": "FE94474 \\"even-keeled\\"",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Feature was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Needs",
                "date": "3:00 AM",
                "documentId": "4bccc111-e5d6-49cb-a5b7-c2c71c88925d",
                "executor": "",
                "message": "Vision-oriented reboot reboot",
              },
              Object {
                "collection": "Needs",
                "date": "3:00 AM",
                "documentId": "4bccc111-e5d6-49cb-a5b7-c2c71c88925d",
                "executor": "",
                "message": "Latvian Lats",
              },
            ],
            "title": "NE77261 \\"Refined Cotton Table\\"",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Need was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Wants",
                "date": "3:00 AM",
                "documentId": "c8df8e65-10df-4065-a5de-a75ec18d9968",
                "executor": "",
                "message": "system",
              },
              Object {
                "collection": "Wants",
                "date": "3:00 AM",
                "documentId": "c8df8e65-10df-4065-a5de-a75ec18d9968",
                "executor": "",
                "message": "orchid",
              },
            ],
            "title": "WA92244 \\"Credit Card Account\\"",
            "url": "99173/canvas",
          },
        ],
        "title": "1 Want was updated in the business model canvas:",
      },
      Object {
        "data": Array [
          Object {
            "logs": Array [
              Object {
                "collection": "Organizations",
                "date": "3:00 AM",
                "documentId": "280cf0c1-71a0-44e6-8df9-b84dfdd0262e",
                "executor": "",
                "message": "Kyat",
              },
            ],
          },
        ],
        "title": "Organization settings were updated:",
      },
    ],
    "title": "Daily recap for January 01, 2018",
    "unsubscribeUrl": "99173/unsubscribe",
  },
  "templateName": "recapEmail",
}
`;
